src/Tools/Source/RunTests/Program.cs

```diff
@@ -107,6 +147,76 @@ private static void WriteLogFile(Options options)
         }
 
         /// <summary>
+        /// Invoked when a timeout occurs and we need to dump all of the test processes and shut down 
+        /// the runnner.
+        /// </summary>
+        private static void HandleTimeout(Options options)
+        {
+            string GetProcDumpFilePath(Process proc) => ProcessUtil.Is64Bit(proc)
+                ? Path.Combine(options.ProcDumpPath, "procdump64.exe")
+                : Path.Combine(options.ProcDumpPath, "procdump.exe");
+
+            void DumpProcess(Process targetProcess, string dumpFilePath)
+            {
+                Console.Write($"Dumping {targetProcess.ProcessName} {targetProcess.Id} to {dumpFilePath} ... ");
+                try
+                {
+                    var processStartInfo = new ProcessStartInfo();
+                    processStartInfo.FileName = GetProcDumpFilePath(targetProcess);
+                    processStartInfo.Arguments = $"-accepteula -ma {targetProcess.Id} {dumpFilePath}";
+                    processStartInfo.CreateNoWindow = true;
+                    processStartInfo.UseShellExecute = false;
+                    processStartInfo.WindowStyle = ProcessWindowStyle.Hidden;
+                    processStartInfo.RedirectStandardOutput = true;
+                    var process = Process.Start(processStartInfo);
+                    process.WaitForExit();
+
+                    // The exit code for procdump doesn't obey standard windows rules.  It will return non-zero
+                    // for succesful cases (possibly returning the count of dumps that were written).  Best 
+                    // backup is to test for the dump file being present.
+                    if (File.Exists(dumpFilePath))
+                    {
+                        Console.WriteLine("succeeded");
+                    }
+                    else
+                    {
+                        Console.WriteLine($"FAILED with {process.ExitCode}");
+                        Console.WriteLine($"{processStartInfo.FileName} {processStartInfo.Arguments}");
+                        Console.WriteLine(process.StandardOutput.ReadToEnd());
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine("FAILED");
+                    Console.WriteLine(ex.Message);
+                }
+            }
+
+            Console.WriteLine("Roslyn Error: test timeout exceeded, dumping remaining processes");
+
+            if (!string.IsNullOrEmpty(options.ProcDumpPath))
+            {
+                var dumpDir = options.LogFilePath != null
+                    ? Path.GetDirectoryName(options.LogFilePath)
+                    : Directory.GetCurrentDirectory();
```

Comment:
Test assembly path isn't a consideration here.  This code will create dumps not only for xunit, but for any child process: devenv, csc, etc ...  Hence we can't bank on the assembly path being available. 

Yes I generally hate working directory but I don't feel like I have a lot of choice here.  Lacking a logging directory there is no other great directory to choose.

I also feel a central directory is preferrable here.  It ptus all the dumps in a single location.  Much better for looking at all of the problems, Jenkins UI (less relevant because we always provide a logpath there).